File: web/lib/contracts.ts
/**
 * TypeScript mirror of the Python application-layer wire contract
 * (`src/adaptive_disclosure_gateway/application/wire.py`) and the request
 * body it accepts (`src/adaptive_disclosure_gateway/api/schemas.py`'s
 * `DisclosureRequestBody`).
 *
 * This module declares TYPES ONLY, plus `CONTRACT_VERSION` and the two
 * closed-set constants the runtime guards in `lib/responseGuards.ts` need
 * (`KNOWN_DISCLOSURE_OUTCOMES`, `DISCLOSURE_SUMMARY_STATUSES`) -- no logic,
 * no mapping, no defaults. Field names and shapes mirror the
 * Python side exactly (snake_case, verbatim), so a response decoded as
 * `JSON.parse` matches these types without any renaming step. See
 * `docs/advisor-demo.md` / CLAUDE.md: the UI never reimplements scientific
 * or security semantics -- it renders what the API returns, so its type
 * layer must not silently diverge from what the API actually sends.
 *
 * `CONTRACT_VERSION` must stay byte-identical to the Python constant of the
 * same name in `application/wire.py` -- it is a published contract
 * identifier, not a version to bump opportunistically.
 */

export const CONTRACT_VERSION = "t20-application-api-v1";

// --- GET /health -----------------------------------------------------------

export interface ProviderHealth {
  provider_class: string;
  model_id: string | null;
  model_snapshot: string | null;
  deterministic_demo_mode: boolean;
}

export interface HealthResponse {
  status: string;
  contract_version: string;
  provider: ProviderHealth;
  treatments_available: string[];
}

// --- GET /examples -----------------------------------------------------------

export interface ExampleSummary {
  example_id: string;
  title: string;
  domain: string;
  purpose: string;
  task: string;
  character_count: number;
}

export interface ExamplesResponse {
  contract_version: string;
  examples: ExampleSummary[];
}

// --- GET /documents/types ---------------------------------------------------

export interface DocumentType {
  document_type: string;
  analysis_modes: string[];
  default_analysis_mode: string;
}

export interface DocumentTypesResponse {
  contract_version: string;
  document_types: DocumentType[];
}

// --- shared preview/execute pieces ------------------------------------------

/**
 * The `DisclosureOutcome` StrEnum values as declared today in
 * `application/contracts.py`. Kept as a plain readonly array (not just a
 * union) so `contracts.test.ts` can diff it against the Python source file
 * on disk -- see that test for the enforcement this backs.
 */
export const KNOWN_DISCLOSURE_OUTCOMES = [
  "removed",
  "pseudonymized",
  "generalized",
  "preserved",
  "blocked",
] as const;

export type KnownDisclosureOutcome = (typeof KNOWN_DISCLOSURE_OUTCOMES)[number];

/**
 * `outcome` is modeled as the known union PLUS an escape hatch
 * (`string & {}`) rather than a bare union or a bare `string`.
 *
 * Why not a bare union: the API is a separate, independently-evolving
 * service (T20). A future outcome added on the Python side must still be
 * *representable* by this type -- otherwise decoding a real response would
 * require an unsound cast (`as DisclosureOutcome`) at the JSON boundary,
 * which defeats the point of typing this at all.
 *
 * Why not a bare `string`: that would erase the known members entirely and
 * silence every switch/lookup exhaustiveness signal in `outcomes.ts`.
 *
 * `string & {}` is a widened-string trick: TypeScript keeps the literal
 * members as the *preferred* completions/narrowing target while still
 * accepting any other string, without collapsing the whole type to `string`
 * (which `KnownDisclosureOutcome | string` would do). This is exactly the
 * shape the fail-closed mapping in `outcomes.ts` depends on: it can switch
 * on the known members and must have a real "otherwise" branch for
 * anything else, enforced by `outcomes.test.ts`.
 */
export type DisclosureOutcome = KnownDisclosureOutcome | (string & {});

export interface CategoryDisclosureSummary {
  category: string;
  outcome: DisclosureOutcome;
  action: string;
  crosses_trust_boundary: boolean;
  occurrence_count: number;
  required_for_task: boolean | null;
  technical_reason: string;
  policy_version: string | null;
  policy_restricted: boolean | null;
  impossible_under_policy: boolean | null;
}

/**
 * The `DisclosureSummary.status` values, as `application/contracts.py`
 * declares them today (`Literal["allowed", "blocked"]`).
 *
 * A readonly array rather than a bare union for the same reason
 * `KNOWN_DISCLOSURE_OUTCOMES` is one: `lib/responseGuards.ts` has to check
 * this set at RUNTIME, and `contracts.test.ts` diffs it against the Python
 * source on disk so the two cannot drift.
 *
 * Unlike `outcome`, this set has no escape hatch for an unrecognized value.
 * That is deliberate and is the opposite call from `DisclosureOutcome`
 * above: an unknown outcome degrades safely to "unknown -- verify", but
 * `status` is what decides whether a send is permitted at all, and there is
 * no third rendering between "may send" and "must not send". A status this
 * UI does not know is therefore a rejected response, not a degraded one.
 */
export const DISCLOSURE_SUMMARY_STATUSES = ["allowed", "blocked"] as const;

export type DisclosureSummaryStatus = (typeof DISCLOSURE_SUMMARY_STATUSES)[number];

export interface DisclosureSummary {
  status: DisclosureSummaryStatus;
  categories: CategoryDisclosureSummary[];
  detected_span_count: number;
  detected_categories: string[];
}

export interface SafeGovernanceView {
  domain: string;
  purpose: string;
  policy_version: string;
  provider_class: string;
  requester_role: string | null;
  requested_pseudonym_scope: string;
}

export interface ProviderMode {
  provider_class: string;
}

// --- POST /disclosure/preview -----------------------------------------------

export interface PreviewResponse {
  contract_version: string;
  summary: DisclosureSummary;
  external_payload: string;
  payload_byte_count: number;
  treatment: string;
  strategy: string;
  governance: SafeGovernanceView;
  provider_mode: ProviderMode;
  /**
   * The T27 / issue #69 visual diff/inspector projection. `null` whenever
   * the deployment has `ADG_ENABLE_DEMO_TRANSPARENCY` off (the historical,
   * pre-T27 behavior for every existing caller) -- never absent, always
   * either `null` or a full `DisclosureInspection`.
   */
  inspection: DisclosureInspection | null;
}

/** Structured-upload preview: the normal safe preview plus an opaque proof. */
export interface DocumentPreviewResponse extends PreviewResponse {
  confirmation_token: string;
}

// --- T27 / issue #69: visual diff / transformation inspector ----------------

/**
 * The `DisclosureAction` (`domain.py`) string values that can appear on an
 * `InspectionSegment.action` for a segment that was actually transformed --
 * i.e. the per-span actions a treatment can apply to a detected value.
 * `BLOCK_REQUEST` and `TASK_DEPENDENT` are deliberately excluded: neither is
 * ever the `action` of an individual `Transformation` the inspector
 * projects (`application/inspection.py`) -- `block_request` stops the whole
 * request before any segment exists, and `task_dependent` is resolved to one
 * of the four members below before a `Transformation` is built. Kept as a
 * plain readonly array (not just a union) so `contracts.test.ts` can pin it
 * against `domain.py`'s `DisclosureAction` on disk, and so
 * `lib/inspectionActions.ts` can check membership at runtime.
 */
export const KNOWN_INSPECTION_ACTIONS = ["preserve", "pseudonymize", "generalize", "remove"] as const;

export type KnownInspectionAction = (typeof KNOWN_INSPECTION_ACTIONS)[number];

/**
 * `action` is modeled the same way `DisclosureOutcome` is above: the known
 * union plus an escape hatch, never a bare union or a bare `string` -- see
 * that type's docstring for why. `null` (a separate case, not part of this
 * union) means the segment was untouched, never an unrecognized action.
 */
export type InspectionAction = KnownInspectionAction | (string & {});

/**
 * One segment of the T27 visual diff/inspector projection
 * (`application/contracts.py`'s `DisclosureInspectionSegment` /
 * `application/wire.py`'s `InspectionSegmentModel`). `action`/`category` are
 * both `null` for an untouched segment -- never independently null/non-null,
 * see `lib/responseGuards.ts`'s invariant check. Segments carry no offsets
 * by design: Python code-point offsets are not JS UTF-16 indices, so the UI
 * must never slice `original`/`disclosed` by any index derived from the API
 * -- it only ever renders these strings and concatenates them in order.
 */
export interface InspectionSegment {
  action: InspectionAction | null;
  category: string | null;
  original: string;
  disclosed: string;
}

/**
 * The T27 visual diff/inspector projection itself
 * (`application/contracts.py`'s `DisclosureInspection` /
 * `application/wire.py`'s `DisclosureInspectionModel`). `segments` is always
 * empty when `available` is `false`. `unavailable_reason` is one of
 * `"blocked"` (the disclosure decision itself blocked the request -- nothing
 * to inspect) or `"alignment_failed"` (the pipeline's own alignment check
 * could not verify the projection is faithful, so it fails closed rather
 * than show a partial/best-effort diff) when `available` is `false`, and
 * `null` when `available` is `true`.
 */
export interface DisclosureInspection {
  available: boolean;
  unavailable_reason: string | null;
  segments: InspectionSegment[];
}

// --- POST /disclosure/execute -----------------------------------------------

export interface ProviderStage {
  called: boolean;
  provider_class: string | null;
  model_id: string | null;
  model_snapshot: string | null;
  decoding_config: Record<string, unknown> | null;
  transmitted_bytes: number | null;
  response_hash: string | null;
  failed: boolean;
  failure_kind: string | null;
}

export interface ReconstructionStage {
  attempted: boolean;
  reconstructed_hash: string | null;
  changed_from_provider_response: boolean | null;
}

export interface ExecuteResponse {
  contract_version: string;
  status: string;
  summary: DisclosureSummary;
  final_answer: string | null;
  provider: ProviderStage;
  reconstruction: ReconstructionStage;
  treatment: string;
  strategy: string;
  governance: SafeGovernanceView;
  total_ms: number;
}

// --- POST /disclosure/compare -----------------------------------------------

/**
 * `application/contracts.py`'s `CANONICAL_COMPARISON_ORDER`, mirrored as the
 * frozen `b0`-`b4` codes in the exact order `service.compare_strategies`
 * iterates it -- DIRECT, STATIC_SANITIZATION, REVERSIBLE_PSEUDONYMIZATION,
 * TASK_AWARE, POLICY_GOVERNED. `contracts.test.ts` diffs this against the
 * Python source on disk. This is presentation order only: the UI never
 * re-sorts `CompareResponse.entries` by this array -- the API already
 * returns them in this order, and re-sorting here would be exactly the
 * kind of "UI reimplements semantics" CLAUDE.md forbids. It exists so
 * per-strategy copy (`copy.treatments`) can be looked up/iterated in a
 * fixed, tested order for things like a legend, independent of what any
 * given response happens to contain.
 */
export const CANONICAL_COMPARISON_ORDER = ["b0", "b1", "b2", "b3", "b4"] as const;

export type ComparisonStrategyCode = (typeof CANONICAL_COMPARISON_ORDER)[number];

/**
 * The treatment code the real wire emits for each position in
 * `CANONICAL_COMPARISON_ORDER`. `StrategyComparisonEntryModel.from_domain`
 * (`application/wire.py`) serializes `treatment=entry.treatment.value`,
 * and `application/contracts.py`'s `resolve_treatment` (backed by
 * `_STRATEGY_TO_TREATMENT`) maps every one of the five canonical
 * `DisclosureStrategy` values to the `Treatment` of the IDENTICAL `b0`-`b4`
 * code -- `DisclosureStrategy.DIRECT` ("b0") to `Treatment.DIRECT` ("b0"),
 * and so on through `POLICY_GOVERNED`/"b4". So today this equals
 * `CANONICAL_COMPARISON_ORDER` verbatim.
 *
 * That identity is verified against the Python source, not assumed: kept
 * as its own named constant (rather than reusing
 * `CANONICAL_COMPARISON_ORDER` again at each call site) so the assumption
 * is visible by name, and `contracts.test.ts` derives this same
 * strategy-to-treatment-code mapping independently from
 * `resolve_treatment`/`_STRATEGY_TO_TREATMENT` in `application/contracts.py`
 * and `Treatment` in `domain.py`, diffing it against this constant -- so if
 * a future change ever made a canonical strategy resolve to a
 * differently-coded treatment, that drift test fails instead of the
 * runtime guard in `responseGuards.ts` silently accepting or rejecting the
 * wrong thing.
 */
export const CANONICAL_COMPARISON_TREATMENTS: readonly ComparisonStrategyCode[] = CANONICAL_COMPARISON_ORDER;

/**
 * One strategy's entry in a `/disclosure/compare` response. `strategy` is
 * always an explicit `b0`-`b4` code, never `"recommended"` (see
 * `application/contracts.py`'s `StrategyComparisonEntry` docstring), and
 * `treatment` is always the `b0`-`b4` code of the treatment that strategy
 * resolved to (see `CANONICAL_COMPARISON_TREATMENTS` above) -- so both
 * fields are narrowed to `ComparisonStrategyCode` rather than left as bare
 * `string`, unlike the other identifier fields this module mirrors
 * verbatim. This is what lets `responseGuards.ts` and its fixtures be
 * checked by the type system, not just at runtime: a fixture assigning
 * `treatment: "policy_governed"` (a human-readable name, never a real wire
 * value) fails to compile instead of only failing a runtime guard test.
 *
 * `unsafe_control_baseline` is the ONLY field the UI may use to detect the
 * B0 -- Direct control: never `strategy === "b0"`. It is derived
 * server-side from the treatment's own capability marker
 * (`pipeline.UnsafeControlTreatment`), not from a hardcoded identifier
 * comparison, and the UI must not reintroduce that hardcoding on its own
 * side either -- see `outcomes.test.ts`-style pin in
 * `ComparisonScreen.test.tsx`.
 */
export interface StrategyComparisonEntry {
  strategy: ComparisonStrategyCode;
  treatment: ComparisonStrategyCode;
  recommended: boolean;
  unsafe_control_baseline: boolean;
  summary: DisclosureSummary;
  external_payload: string;
  payload_byte_count: number;
}

export interface CompareResponse {
  contract_version: string;
  entries: StrategyComparisonEntry[];
  governance: SafeGovernanceView;
  provider_mode: ProviderMode;
}

// --- error bodies ------------------------------------------------------------

/** The safe body for every non-validation error response (400/404/500). */
export interface ErrorResponse {
  detail: string;
  kind: string;
}

/** One pydantic validation error, stripped of FastAPI's default `input` echo. */
export interface ValidationErrorItem {
  loc: string[];
  type: string;
  msg: string;
}

/** The 422 body shape returned by FastAPI's `RequestValidationError` handler. */
export interface ValidationErrorResponse {
  detail: ValidationErrorItem[];
}

// --- export / restore (T26 / issue #67; gated behind -----------------------
// ADG_ENABLE_DEMO_TRANSPARENCY for the web UI -- T28 / #70). This module
// never spells out either route handler's own upstream path string in
// prose -- see tests/test_demo_deployment_config.py::TestWebExportRestoreIsGated,
// which pins those two literal strings to the route handlers and lib/api.ts
// only.

/**
 * `application/wire.py`'s `ExportResponse`. Deliberately excludes anything
 * that would let the pseudonym -> original mapping travel wholesale: only
 * `restore_handle` (opaque) and `restorable_count`, never the mapping
 * entries themselves. `expires_at` is an epoch-seconds integer, not an ISO
 * string -- the UI formats it for display, never re-derives a security
 * decision from it.
 */
export interface ExportResponse {
  contract_version: string;
  external_payload: string;
  restore_handle: string;
  expires_at: number;
  restorable_count: number;
  treatment: string;
  strategy: string;
  governance: SafeGovernanceView;
}

/**
 * `application/wire.py`'s `RestoreResponse`. Never carries the mapping
 * either -- only the restored text and two counts.
 */
export interface RestoreResponse {
  contract_version: string;
  restored_text: string;
  restored_count: number;
  unresolved_count: number;
}

// --- GET /api/demo/features (web-only; not part of application/wire.py) -----

/**
 * A web-invented, purely presentational endpoint (T28 / issue #70): whether
 * the export/restore UI should render at all, computed server-side from the
 * SAME `ADG_ENABLE_DEMO_TRANSPARENCY` gate the route handlers themselves
 * enforce (`lib/demoTransparency.ts`). Carries no `contract_version` -- it
 * is not part of the Python application-layer wire contract this module
 * otherwise mirrors, and never will be: it exists only so the client knows
 * whether to bother rendering export/restore controls, never to decide
 * anything security-relevant (the route handlers gate that independently,
 * per request, regardless of what this endpoint last reported).
 */
export interface DemoFeaturesResponse {
  demo_transparency_enabled: boolean;
}

// --- request body ------------------------------------------------------------

/** Mirrors `api/schemas.py`'s `GovernanceOverridesBody` field for field. */
export interface GovernanceOverridesBody {
  purpose?: string | null;
  requester_role?: string | null;
  requester_id?: string | null;
  provider_class?: string | null;
  policy_version?: string | null;
  requested_pseudonym_scope?: string | null;
  session_id?: string | null;
  document_id?: string | null;
  request_id?: string | null;
  domain?: string | null;
}

/**
 * The one shared request body for `POST /disclosure/preview` and
 * `POST /disclosure/execute`. `strategy` mirrors
 * `application.contracts.DisclosureStrategy` ("recommended" | "b0".."b4"),
 * kept as a plain string here since this module is types-only.
 */
export interface DisclosureRequestBody {
  text?: string | null;
  filename?: string | null;
  file_content?: string | null;
  example_id?: string | null;
  task?: string | null;
  strategy?: string | null;
  governance?: GovernanceOverridesBody | null;
}

File: src/adaptive_disclosure_gateway/application/contracts.py
"""The stable application-layer contract for the disclosure preview/execute
use cases (T20 / issue #28, slice 1).

Every result contract here is a frozen dataclass, and every one of them is
no-leak safe by construction:

- ``Transformation.original``/``Transformation.transformed`` (the vault
  mapping) never appear anywhere below -- only counts derived from them
  (``CategoryDisclosureSummary.occurrence_count``, see ``summaries.py``).
- ``DisclosurePreview.external_payload`` IS returned deliberately: it is
  exactly the representation that would cross the trust boundary, returned
  to the same caller who supplied the source content, and it is the whole
  point of a "review before sending" screen. This is not a leak -- it is the
  product surface this slice exists to build.
- ``SafeGovernanceView`` exposes only ``domain``/``purpose``/``policy_version``/
  ``provider_class``/``requester_role``/``requested_pseudonym_scope``. It
  never exposes ``requester_id`` or any lifecycle identifier
  (``session_id``/``document_id``/``request_id``) -- see
  ``tests/test_application_contracts.py``'s dedicated pin.
- ``DisclosureExecution.provider``/``.reconstruction`` reuse
  ``audit.ProviderStage``/``audit.ReconstructionStage`` verbatim rather than
  redefining an equivalent shape: those types are already no-leak safe by
  construction (metadata/hashes only, never a raw value or provider message
  -- see ``audit.py``'s own module docstring), and redefining them here
  would be exactly the kind of duplicated-shape drift this ticket's "no
  duplicated logic" rule warns against.

``DisclosureExecution.total_ms`` is a basic wall-clock convenience for the
UI (how long the whole use case took, start to finish). It is explicitly
NOT the T10 scientific latency metric (``experiments/stage_timing.py``,
which requires the exact span-topology proof T10 relies on) and must never
be used as one.

``StrategyComparisonEntry``/``StrategyComparison`` (T20 / issue #28's
"Compare strategies" slice, issue #29/#41): the educational surface that
runs the SAME content through every B0-B4 strategy and shows what each one
would disclose.

This comparison is PREVIEW-based and must NEVER call a provider, for any
strategy -- see ``service.compare_strategies``'s own docstring for the full
reasoning (in short: B0 -- Direct discloses the raw document unchanged by
design, so *executing* a comparison would send the caller's unprotected
document to an external provider merely to illustrate a teaching point;
with ``FakeProvider`` the responses carry no task utility at all today, so
executing adds nothing now and only creates that risk once T22/#30 wires in
a real provider). Showing exactly what each strategy *would* send is the
entire pedagogical content of the comparison. An execute-based/
utility-aware comparison is explicitly deferred, and would require both
T22 and a deliberate decision about whether B0 may ever run against a real
provider on user content.

This is also NOT an evaluation surface: it never imports
``experiments.scoring``, never touches the oracle, never computes
conformance/exposure/unnecessary-disclosure or any T10 metric, and never
ranks/scores/declares a "best" strategy -- it reports facts each preview
already produced. The only judgment it expresses is the pre-existing
``recommended`` flag ``list_strategy_options`` already documents as a
product/UX default, not a scientific claim. No aggregate counts, ratios,
deltas or "protection score" are added here -- a UI that wants those can
already derive them from ``summary.categories`` on each entry.

``StrategyComparisonEntry.strategy`` is always one of the five explicit
``b0``-``b4`` values, never ``DisclosureStrategy.RECOMMENDED`` --
``recommended`` is a separate boolean precisely so ``RECOMMENDED`` (a
sentinel meaning "resolve this for me") never needs to appear as a value
alongside the four explicit strategies it could resolve to.
``unsafe_control_baseline`` is derived from the existing
``pipeline.UnsafeControlTreatment`` capability marker -- never a hardcoded
comparison against ``Treatment.DIRECT`` -- so it stays correct if a future
treatment ever gained the same marker.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import Literal

from adaptive_disclosure_gateway.application.ingestion import NormalizedContent
from adaptive_disclosure_gateway.audit import ProviderStage, ReconstructionStage
from adaptive_disclosure_gateway.domain import (
    DisclosureAction,
    GovernanceContext,
    PseudonymScope,
    Treatment,
)


class DisclosureStrategy(StrEnum):
    """What the caller asks for. ``RECOMMENDED`` is the novice default and
    is the only value the guided UI needs to know; the explicit B0-B4
    values exist for the optional research/comparison surface.

    Values reuse ``Treatment``'s own frozen ``"b0"``-``"b4"`` codes verbatim
    (CLAUDE.md: frozen codes must never change) -- no new codes are minted
    here.
    """

    RECOMMENDED = "recommended"
    DIRECT = "b0"
    STATIC_SANITIZATION = "b1"
    REVERSIBLE_PSEUDONYMIZATION = "b2"
    TASK_AWARE = "b3"
    POLICY_GOVERNED = "b4"


# RECOMMENDED -> POLICY_GOVERNED is a product/UX default for the guided demo
# -- "the treatment we recommend a novice caller use" -- and is not a
# scientific claim that B4 dominates every comparison; it never changes any
# treatment's semantics, and every explicit B0-B4 value still maps to
# exactly the treatment it names.
_STRATEGY_TO_TREATMENT: dict[DisclosureStrategy, Treatment] = {
    DisclosureStrategy.DIRECT: Treatment.DIRECT,
    DisclosureStrategy.STATIC_SANITIZATION: Treatment.STATIC_SANITIZATION,
    DisclosureStrategy.REVERSIBLE_PSEUDONYMIZATION: Treatment.REVERSIBLE_PSEUDONYMIZATION,
    DisclosureStrategy.TASK_AWARE: Treatment.TASK_AWARE,
    DisclosureStrategy.POLICY_GOVERNED: Treatment.POLICY_GOVERNED,
    DisclosureStrategy.RECOMMENDED: Treatment.POLICY_GOVERNED,
}


def resolve_treatment(strategy: DisclosureStrategy) -> Treatment:
    """Map a caller-facing ``DisclosureStrategy`` to the ``Treatment`` the
    core actually runs. See the module-level comment above for why
    ``RECOMMENDED`` maps to ``Treatment.POLICY_GOVERNED``.
    """
    return _STRATEGY_TO_TREATMENT[strategy]


# CLAUDE.md's canonical treatment sequence, Direct -> Static Sanitization ->
# Reversible Pseudonymization -> Task-aware -> Policy-governed, expressed as
# an explicit, named tuple of the five explicit DisclosureStrategy values
# (never RECOMMENDED) -- deliberately NOT `tuple(DisclosureStrategy)` (whose
# declaration order happens to put RECOMMENDED first and would silently
# change if that enum's declaration order ever did) and not derived from
# Treatment's own declaration order either. `service.compare_strategies`
# iterates this tuple, and only this tuple, to decide entry order.
CANONICAL_COMPARISON_ORDER: tuple[DisclosureStrategy, ...] = (
    DisclosureStrategy.DIRECT,
    DisclosureStrategy.STATIC_SANITIZATION,
    DisclosureStrategy.REVERSIBLE_PSEUDONYMIZATION,
    DisclosureStrategy.TASK_AWARE,
    DisclosureStrategy.POLICY_GOVERNED,
)


@dataclass(frozen=True)
class StrategyOption:
    """One entry of the B0-B4 + recommended discovery list (T20 / issue #28,
    slice 2's ``GET /strategies``). Pure data -- listing these never
    executes anything, and never carries human prose: the UI owns copy
    (issue #29 -- "no scientific identifiers should be translated
    internally").
    """

    strategy: DisclosureStrategy
    treatment: Treatment
    recommended: bool


def list_strategy_options() -> tuple[StrategyOption, ...]:
    """Every ``DisclosureStrategy`` value paired with the ``Treatment`` it
    resolves to and whether it is the novice default -- in
    ``DisclosureStrategy``'s own declaration order, so ``RECOMMENDED`` is
    always first.
    """
    return tuple(
        StrategyOption(
            strategy=strategy,
            treatment=resolve_treatment(strategy),
            recommended=strategy is DisclosureStrategy.RECOMMENDED,
        )
        for strategy in DisclosureStrategy
    )


@dataclass(frozen=True)
class GovernanceOverrides:
    """Optional per-request overrides for the server-configured default
    ``GovernanceContext``. Every field optional: a novice caller supplies
    none of these and gets the server's configured defaults untouched.
    """

    purpose: str | None = None
    requester_role: str | None = None
    requester_id: str | None = None
    provider_class: str | None = None
    policy_version: str | None = None
    requested_pseudonym_scope: PseudonymScope | None = None
    session_id: str | None = None
    document_id: str | None = None
    request_id: str | None = None
    domain: str | None = None


@dataclass(frozen=True)
class DocumentRequestDescriptor:
    """The caller-facing selection an uploaded document was analysed under.

    Carried on the request so the preview-confirmation fingerprint can bind
    what the *caller asked for* (``document_type``/``analysis_mode``) and not
    only what the server resolved it to. The two are not interchangeable: a
    future preset could map two document types onto one governance
    configuration, and an approval of one must not authorise the other.

    ``analysis_mode`` is always the resolved mode -- never ``None`` -- so a
    preview that accepted the preset's default and an execute that named
    that same mode explicitly describe one state rather than two.
    """

    document_type: str
    analysis_mode: str


@dataclass(frozen=True)
class DisclosureApplicationRequest:
    """The single input shape both ``preview`` and ``execute`` accept.

    ``document`` is set only by ``build_document_request`` (the structured
    upload path) and stays ``None`` for pasted text, ``.txt``/``.md`` files
    and prepared examples. It is what distinguishes a request that must
    carry a preview confirmation from one on the historical ``/disclosure/*``
    surface, which is unchanged by that mechanism.
    """

    content: NormalizedContent
    task: str
    strategy: DisclosureStrategy = DisclosureStrategy.RECOMMENDED
    governance: GovernanceOverrides = field(default_factory=GovernanceOverrides)
    document: DocumentRequestDescriptor | None = None


class DisclosureOutcome(StrEnum):
    """Stable, presentation-facing outcome codes. The UI maps these to
    localized copy; this layer never emits localized prose.
    """

    REMOVED = "removed"
    PSEUDONYMIZED = "pseudonymized"
    GENERALIZED = "generalized"
    PRESERVED = "preserved"
    BLOCKED = "blocked"


@dataclass(frozen=True)
class CategoryDisclosureSummary:
    """One detected category's disclosure outcome for the "what happens to
    my data" screen. Never carries a raw value or pseudonym -- only counts,
    flags and the treatment's own (non-sensitive) reasoning text.
    """

    category: str
    outcome: DisclosureOutcome
    action: DisclosureAction
    crosses_trust_boundary: bool
    occurrence_count: int
    required_for_task: bool | None
    technical_reason: str
    policy_version: str | None
    policy_restricted: bool | None
    impossible_under_policy: bool | None


@dataclass(frozen=True)
class DisclosureSummary:
    """The full "what happens to my data" picture for one request, shared
    verbatim between ``DisclosurePreview`` and ``DisclosureExecution`` --
    see ``summaries.build_disclosure_summary``.
    """

    status: Literal["allowed", "blocked"]
    categories: tuple[CategoryDisclosureSummary, ...]
    detected_span_count: int
    detected_categories: tuple[str, ...]


@dataclass(frozen=True)
class SafeGovernanceView:
    """A ``GovernanceContext`` projected down to what is safe to hand back
    to the same caller who supplied it: domain/purpose/policy_version/
    provider_class/requester_role/pseudonym scope only.

    Deliberately excludes ``requester_id`` and every lifecycle identifier
    (``session_id``/``document_id``/``request_id``): those identify a
    specific requester/session/document and partition the pseudonym vault,
    and have no presentational value to a "what happens to my data" screen
    -- see ``tests/test_application_contracts.py``'s dedicated pin that this
    type has no such field at all, not merely that it is left unset.
    """

    domain: str
    purpose: str
    policy_version: str
    provider_class: str
    requester_role: str | None
    requested_pseudonym_scope: PseudonymScope


def safe_governance_view(context: GovernanceContext) -> SafeGovernanceView:
    """Project a real ``GovernanceContext`` down to ``SafeGovernanceView``."""
    return SafeGovernanceView(
        domain=context.domain,
        purpose=context.purpose,
        policy_version=context.policy_version,
        provider_class=context.provider_class,
        requester_role=context.requester_role,
        requested_pseudonym_scope=context.requested_pseudonym_scope,
    )


@dataclass(frozen=True)
class ProviderMode:
    """Which provider *would* handle a previewed request, known without
    ever calling it -- just the provider's own declared ``provider_class``.
    Populated by ``preview``, which never invokes a provider at all (see
    ``service.py``).
    """

    provider_class: str


@dataclass(frozen=True)
class DisclosurePreview:
    """The "review before sending" result: everything about what would
    happen, plus the exact payload that would cross the trust boundary --
    without ever calling a provider.
    """

    summary: DisclosureSummary
    external_payload: str
    payload_byte_count: int
    treatment: Treatment
    strategy: DisclosureStrategy
    governance: SafeGovernanceView
    provider_mode: ProviderMode


@dataclass(frozen=True)
class DocumentDisclosurePreview:
    """A structured-document preview plus the server-signed proof of what
    was reviewed.

    A separate type rather than an optional field on ``DisclosurePreview``:
    the historical ``/disclosure/preview`` surface does not have (or need) a
    confirmation, and giving it a permanently-null ``confirmation_token``
    would change its contract for every existing caller to describe a
    mechanism that does not apply to it. See
    ``application/preview_confirmation.py``.
    """

    preview: DisclosurePreview
    confirmation_token: str


class UnsafeControlExecutionError(Exception):
    """Raised when an unsafe-control treatment (B0 -- Direct) is asked to
    execute an uploaded document against a provider outside the trust
    boundary.

    A product/demo-surface rule, not an experimental one. B0's experimental
    semantics are untouched and it stays fully visible in preview and in the
    B0-B4 comparison; what it must not do is send an advisor's untransformed
    document across the organizational boundary. This is the same stance
    ``compare_strategies`` already takes when it refuses to execute any
    comparison entry.

    Its message names the treatment class and the surface only -- never the
    document, the task or the payload.
    """


@dataclass(frozen=True)
class DisclosureExecution:
    """The real result of running a request all the way through the core:
    provider call, local reconstruction (if the treatment supports it), and
    the same disclosure summary ``preview`` would have shown for this input.
    """

    summary: DisclosureSummary
    final_answer: str | None
    provider: ProviderStage
    reconstruction: ReconstructionStage
    treatment: Treatment
    strategy: DisclosureStrategy
    governance: SafeGovernanceView
    total_ms: float


class ExportRefusedError(Exception):
    """Raised when ``DisclosureApplicationService.export`` is asked to
    export a request whose disclosure decision is not ``"allowed"`` (T26 /
    issue #67) -- a ``BLOCK_REQUEST`` outcome, or any other case where
    ``preview`` would not have produced a disclosed representation at all.

    Export exists to hand a caller the SAME disclosed representation
    ``preview`` already shows them, plus a restore handle for it -- there is
    nothing to export for a request that never produced one. Its message
    names the surface only, never the document, the task, or which category
    caused the block (that detail already exists, safely, on the ordinary
    preview response).
    """


@dataclass(frozen=True)
class DisclosureExport:
    """The result of ``DisclosureApplicationService.export`` (T26 / issue
    #67): the same disclosed representation ``preview``/``execute`` would
    show, plus a stateless, sealed restore handle for it.

    ``external_payload`` is exactly ``DisclosurePreview.external_payload``
    for the same input -- see that field's own docstring for why returning
    it is the product, not a leak. ``restore_handle`` is an opaque,
    self-contained envelope (see ``application/restore_handle.py``); this
    service never retains anything server-side to make it work later.
    ``restorable_count`` is the number of pseudonym entries the handle
    actually carries -- zero for B0 -- Direct and B1 -- Static Sanitization,
    which never pseudonymize anything, and for any category a treatment
    removed or generalized instead.
    """

    external_payload: str
    restore_handle: str
    expires_at: int
    restorable_count: int
    treatment: Treatment
    strategy: DisclosureStrategy
    governance: SafeGovernanceView


@dataclass(frozen=True)
class DisclosureRestore:
    """The result of ``DisclosureApplicationService.restore`` (T26 / issue
    #67): the submitted text with every pseudonym the handle recognizes
    replaced by its original, plus two counts -- never the mapping, and
    never an original for a pseudonym absent from the submitted text.

    ``unresolved_count`` is the number of pseudonym-shaped tokens present in
    the submitted text that this handle does NOT know about -- e.g. because
    they belong to a different document's export. They are left untouched in
    ``restored_text``, never reported individually.
    """

    restored_text: str
    restored_count: int
    unresolved_count: int


@dataclass(frozen=True)
class StrategyComparisonEntry:
    """One strategy's preview within a ``StrategyComparison`` -- see the
    module docstring's "Compare strategies" section.

    ``strategy`` is always an explicit ``b0``-``b4`` value, never
    ``DisclosureStrategy.RECOMMENDED``. ``recommended`` is ``True`` only for
    the one entry whose ``treatment`` is the treatment
    ``DisclosureStrategy.RECOMMENDED`` currently resolves to.
    ``unsafe_control_baseline`` is ``True`` only for the entry produced by a
    treatment implementing ``pipeline.UnsafeControlTreatment`` -- today only
    B0 -- Direct.
    """

    strategy: DisclosureStrategy
    treatment: Treatment
    recommended: bool
    unsafe_control_baseline: bool
    summary: DisclosureSummary
    external_payload: str
    payload_byte_count: int


@dataclass(frozen=True)
class StrategyComparison:
    """The full "compare strategies" result: one ``StrategyComparisonEntry``
    per B0-B4 strategy, in ``CANONICAL_COMPARISON_ORDER``, plus the
    governance view and provider mode shared identically by every entry
    (content/task/governance are byte-for-byte identical across entries by
    construction -- see ``service.compare_strategies``).
    """

    entries: tuple[StrategyComparisonEntry, ...]
    governance: SafeGovernanceView
    provider_mode: ProviderMode
